python/TestHarness/testers/RunApp.py

```diff
@@ -177,93 +177,69 @@ def getCommand(self, options):
-        else:
-            command = specs['executable'] + timing_string + specs['input_switch'] + ' ' + specs['input'] + ' ' + ' '.join(specs['cli_args'])
 
-        if options.pbs:
-            return self.getPBSCommand(options)
+        elif options.pbs:
+            # Open our pbs_template.i
+            template_script = open(os.path.join(specs['moose_dir'], 'python', 'TestHarness', 'pbs_template.i'), 'r')
+            content = template_script.read()
+            template_script.close()
+
+            # Add threads to cli_args when using PBS
+            if options.parallel or ncpus > 1 or nthreads > 1:
+                specs['cli_args'].append(' --n-threads=' + str(nthreads))
+
+            # Convert MAX_TIME to hours:minutes for walltime use
+            hours = int(int(specs['max_time']) / 3600)
+            minutes = int(int(specs['max_time']) / 60) % 60
+            specs['walltime'] = '{0:02d}'.format(hours) + ':' + '{0:02d}'.format(minutes) + ':00'
+
+            # Truncate JOB_NAME. PBS can only accept 13 character (6 characters from test name + _## (test serial number) + _### (serialized number generated by cluster_launcher) = the 13 character limit)
+            specs['job_name'] = specs['input'][:6] + '_' + str(options.test_serial_number).zfill(2)
+            specs['job_name'] = specs['job_name'].replace('.', '')
+            specs['job_name'] = specs['job_name'].replace('-', '')
+
+            # Convert TEST_NAME to input tests file name (normally just 'tests')
+            specs['no_copy'] = options.input_file_name
+
+            # Are we using the PBS Emulator? Make this param valid if so.
+            # Add the substitution string here so it is not visable to the user
+            specs.addStringSubParam('pbs_stdout', 'PBS_STDOUT', "Save stdout to this location")
+            specs.addStringSubParam('pbs_stderr', 'PBS_STDERR', "Save stderr to this location")
+            if options.PBSEmulator:
+                specs['pbs_stdout'] = 'pbs_stdout = PBS_EMULATOR'
+                specs['pbs_stderr'] = 'pbs_stderr = PBS_EMULATOR'
+            else:
+                # The PBS Emulator fails when using the PROJECT argument (#PBS -P <project name>)
+                specs.addStringSubParam('pbs_project', 'PBS_PROJECT', "Identify this job submission with this project")
+                specs['pbs_project'] = 'pbs_project = %s' % (options.pbs_project)
 
-        return command
 
-    def getPBSCommand(self, options):
-        if options.parallel == None:
-            default_ncpus = 1
-        else:
-            default_ncpus = options.parallel
+            # Squash cli_args list. Needs to happen before we attempt to replace template with valid parameters.
+            specs['cli_args'] = '"' + ' '.join(specs['cli_args']) + '"'
```

Comment:
Getting closer. However we are still doing one or two things in both major branches here. Building the actual command cli_args appears to be one of those things that's duplicated. Can this be captured from the other branch (i.e. factor it out and do cli_args in one place only).